components/eventing-controller/reconciler/subscription-nats/reconciler.go

```diff
@@ -36,7 +74,244 @@ func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	}
+
+	// Check for valid protocol
+	actualProtocol := actualSubscription.Spec.Protocol
+	if err := r.assertProtocolValidity(actualProtocol); err != nil {
+		r.Log.Error(err, "invalid subscription protocol")
+		err := r.syncSubscriptionStatus(ctx, actualSubscription, false, err.Error())
+		if err != nil {
+			r.Log.Error(err, "failed to sync subscription status")
+			return ctrl.Result{}, err
+		}
+		// No point in reconciling as the sink is invalid
+		return ctrl.Result{}, nil
+	}
+
+	// Check for valid sink
+	if err := r.assertSinkValidity(actualSubscription.Spec.Sink); err != nil {
+		r.Log.Error(err, "failed to parse sink URL")
+		err := r.syncSubscriptionStatus(ctx, actualSubscription, false, err.Error())
+		if err != nil {
+			r.Log.Error(err, "failed to sync subscription status")
+			return ctrl.Result{}, err
+		}
+		// No point in reconciling as the sink is invalid
+		return ctrl.Result{}, nil
+	}
+
+	// The object is not being deleted, so if it does not have our finalizer,
+	// then lets add the finalizer and update the object. This is equivalent
+	// registering our finalizer.
+	if !utils.ContainsString(desiredSubscription.ObjectMeta.Finalizers, Finalizer) {
+		log.Info("Adding finalizer to subscription object")
+		desiredSubscription.ObjectMeta.Finalizers = append(desiredSubscription.ObjectMeta.Finalizers, Finalizer)
+		if err := r.Update(context.Background(), desiredSubscription); err != nil {
+			return ctrl.Result{}, err
+		}
+		result.Requeue = true
+	}
 
-	return ctrl.Result{}, nil
+	if result.Requeue {
+		return result, nil
+	}
+
+	var filters []*eventingv1alpha1.BebFilter
+	if desiredSubscription.Spec.Filter != nil {
+		filters = desiredSubscription.Spec.Filter.Filters
+	}
+
+	// Create subscriptions in Nats
+	for _, filter := range filters {
+		eventType := filter.EventType.Value
+		callback := func(msg *nats.Msg) {
+			ce, err := convertMsgToCE(msg)
+			if err != nil {
+				r.Log.Error(err, "failed to convert Nats message to CE")
+				return
+			}
+			ctx := cloudevents.ContextWithTarget(context.Background(), desiredSubscription.Spec.Sink)
+			cev2Client, err := ceclient.NewDefault()
+			if err != nil {
+				r.Log.Error(err, "failed to create CE client")
+				return
+			}
+			err = cev2Client.Send(ctx, *ce)
+			if err != nil {
+				if !strings.Contains(err.Error(), "200") {
+					log.Error(err, "failed to dispatch event")
+					return
+				}
+			}
+			r.Log.Info(fmt.Sprintf("Successfully dispatched event id: %s to sink: %s", ce.ID(),
+				desiredSubscription.Spec.Sink))
+		}
+
+		if r.natsClient.Connection.Status() != nats.CONNECTED {
+			r.Log.Info("connection to Nats", "status",
+				fmt.Sprintf("%v", r.natsClient.Connection.Status()))
+			initializeErr := r.natsClient.Initialize(r.config)
+			if initializeErr != nil {
+				log.Error(initializeErr, "failed to reset NATS connection")
+				return result, initializeErr
+			}
+		}
+
+		sub, subscribeErr := r.natsClient.Connection.Subscribe(eventType, callback)
```

Comment:
The "eventType" is mapped here directly to a NATS "subject". The convention to map a CE to a NATS "subject" should be: CE.eventSource + CE.eventType

Reply:
This is a nice-to-have in my opinion as eventType is a unique Nats subject already. We can communicate with the user through documentation to use a source that is following the BEB source protocol. This will enable easy portability of subscriptions from open source to other environments.

Reply:
OK, then the convention for the "publisher to NATS"  would be to use only "CE.eventType" from its CE message (ignoring "CE.eventSource") to determine the "subject" where it should send the message.

Reply:
No, CE-source is already translated into the eventtype by the event-publisher proxy(that's how BEB publisher proxy behave and NATS publisher should follow the same). This is done by creating a new event-type as `prefix.<ce-source>.<ce-type>.version` E.g. `sap.kyma.custom.commerce.order.created.v1`.
Here, prefix is `sap.kyma.custom`
ce-source is `commerce`
ce-type is `order.created`
version is `v1`